qos: make sure to wait for service level updates on shutdown

psarna opened the pull request:
The service level controller spawns an updating thread,
which wasn't properly waited for during shutdown.
This behavior is now fixed.

Tests: manual

Fixes #8468

nyh commented on service/qos/service_level_controller.cc:
```diff
@@ -87,14 +88,21 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
+        }
+        co_await std::exchange(_distributed_data_updater, make_ready_future<>());
+    } catch (const broken_semaphore&) {
+        // semaphore errors are expected
```
A question:
If the first co_await throws, the second co_await is not done at all. Is this fine or impossible (i.e. the first one doesn't throw)?
If it isn't fine, you need to use one of the idioms that waits for multiple futures, and probably not coroutines :-(

psarna replied:
If the first co_await throws, then the second one is already ready in this specific situation, so it's safe to discard. And if the second one failed, the first one would never be assigned to, so it will be a ready future and it will not throw.

nyh replied:
Ok, I'll take your word for it - I don't understand when are these broken semaphore et al. exceptions are expected, and why it's enough to only wait for one of these futures in that case, but you obviously know the flow and are correct.

bhalevy replied:
For the sake of being future-proof, I tend to agree with @nyh that we better `co_await when_all_succeed()`
on both, if `_distributed_data_updater` should be ready in case `_global_controller_db->distributed_data_update` failed.

psarna replied:
sure, I'll rework and send v2

nyh commented:
I see you specifically asked Benny for a review, so I won't commit yet.

bhalevy commented:
> I see you specifically asked Benny for a review, so I won't commit yet.

I don't have any clear objection, but I agree with you it's safer on the long run to always wait on both futures in the `this_shard_id() == global_controller` case, even if in the current state of affairs we don't have to.  It's just the the starts alignment might change over time and the proposed change is quite fragile in this sense.

psarna commented:
v2:
 - reversed the order of waiting to make it clearer - we first ensure that initialization of the data updater is finished, and only then wait for the actual updates to end
 - made the steps more robust wrt. exceptions - whatever happens, both routines will always be tried

bhalevy reviewed (approved):
LGTM, thanks.

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -89,12 +90,21 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
+            co_await std::exchange(_global_controller_db->distributed_data_update, make_ready_future<>());
+        } catch (const broken_semaphore& ignored) {
+        } catch (const sleep_aborted& ignored) {
+        }
```
There's no point in having stop() functions pass an exception. So better to catch ... and log it here rather than confusing the caller.

psarna replied:
Yeah, but unexpected exceptions in stop() are caught in main.cc and printed - here I just explicitly ignore the boring ones.

bhalevy replied:
You can still catch the unexpected ones and log an error or warning about them.
An ERROR would help catch it in dtest and decide what do about it since we fail all dtests on any unexpected ERROR that is not explicitly marked to be ignored by the test, while it won't cause a coredump when shutting down in production.

avikivity commented:
Looks good, I'll let @nyh commit.

nyh commented:
> Looks good, I'll let @nyh commit.

I already did. It hasn't yet reached master, so this PR is not closed yet.

psarna commented:
Dequeued due to dtests failures (unexpected error logs). The cause is known so I'll push an updated version in a few minutes.

psarna commented:
Fixed by no bothering to catch only certain types of exceptions after all - not worth all the failed dtests. Please rereview.

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -89,12 +90,21 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
+    try {
+        co_await std::exchange(_distributed_data_updater, make_ready_future<>());
+    } catch (...) {
+        // ignore all exceptions from distributed updater initialization
```
Maybe log at least? we could have something broken and not know about it.

bhalevy replied:
Yeah, I had the same thing in my mind, though I don't how helpful it will be...
For example: https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/3405/artifact/logs-gating.release.2/1618393293064_deletion_test.RangeDeletionTester.delete_by_1ck_range_less_more_test/node4.log
```
ERROR 2021-04-14 09:41:32,078 [shard 0] init - Unexpected error shutting down service level controller: exceptions::unavailable_exception (Cannot achieve consistency level for cl ONE. Requires 1, alive 0)
```
It won't hurt to WARN about it though.

psarna replied:
In that case I'd rather go back to catching specific exception types - `exceptions::unavailable_exception` was the one that got detected in dtests. Listing all expected exception types is the best way to ensure that we don't ignore something genuinely broken, even though it might involve a couple of failed dtest runs to recognize all the expected failures.

I'll restore catching specific types in v3 and add the newly detected error.

bhalevy replied:
why not log a warning about ignoring the exception here?

psarna replied:
If the intent is that we want to know if an unexpected exception happens due to a bug, then it's best to just use the existing mechanism - report an unexpected exception on shutdown as an error and investigate. I don't remember how dtests react to warnings, but I think it's not a good idea either way:
 - if warnings also fail the test, then there's no point in having a separate mechanism for reporting it, main.cc already does that
 - if warnings don't fail the test, then we actually wouldn't know when a bug happens, because nobody reads logs from successful dtests

bhalevy replied:
> If the intent is that we want to know if an unexpected exception happens due to a bug

But this hunk swallows *all* exceptions, with no indication whatsoever that they happen.
I proposed a warning here since it doesn't cause tests to fail but could come handy in postmortem analysis as extra evidence to something else that went bad.

psarna replied:
Ooh we're talking about the first one. Sorry, sure, I forgot it existed. Warning makes perfect sense here.

psarna commented:
Actually, I think it's also possible to race between setting the `_sl_data_accessor` to nullptr and accessing it, because they happen between preemption points. So I'll include both fixes in v3.

psarna commented:
v3:
 - instead of ignoring all exceptions (v2), let's just ignore the one that caused the regression
 - moved setting `_sl_data_accessor` to nullptr only after the updates are done - otherwise it's possible to nullify the pointer before it's being dereferenced by an already ongoing update process

nyh commented:
> Fixed by not bothering to catch only certain types of exceptions after all - not worth all the failed dtests. Please rereview.

Can you give an example what kind of error this was?
Maybe we should consider changing the dtests to be more forgiving for this kind of error? It's very frustrating when dtests enshrine some behavior that we never intended to enshrine.

psarna commented:
@nyh It was an unavailable exception (a CQL one) coming from the fact that the updater service tried to issue a CQL request and failed to perform it, because the cluster is shutting down. It's definitely one of the expected errors here.

nyh reviewed (approved):
I'll commit because you clearly have given this a lot of thought, but I have to be honest - I don't understand the logic where for one future you ignore all exceptions, but for another future you ignore only specific kinds of exceptions. I don't know how to check if this list is comprehensive or not. Though I guess we can incrementally improve it later....

Regarding whether stop() should sometimes throw or just low internal exceptions, see also https://github.com/scylladb/seastar/issues/724 where I debated myself whether a close() should throw, or should catch all internal exceptions and just log them.

bhalevy commented:
Regarding the dtests, by default they want to fail on any unexpected ERROR.  Expected errors can be explicitly added to `ignore_log_pattern` that filters the logs for expected error upon exit.

We can even tolerate expected core dumps when we know the test may crash scylla (like when running the scrub compaction tests that may cause crashes with random sstable injected corruption), but in general I'd rater dtests always fail if a core dump is detected.

nyh commented:
> Regarding the dtests, by default they want to fail on any unexpected ERROR. Expected errors can be explicitly added to `ignore_log_pattern` that filters the logs for expected error upon exit.

Yes, I know this is the case, but I wonder about the wisdom of doing this.
Some tests are clearly about testing **functionality** of something. You test some query and see if you get the expected result. If you get the expected result, and **also** some error is logged, well, it's a shame, but it's not really related to what was being tested. This is doubly true if it's not really a new error but just some old error that was previously ignored but the message changed. It is triply true when the error happens during the shutdown of Scylla - not during the main part of the test.

The most important question in this regard of these error-message-failures, how many were real bugs detected, and how many were false positives that caused any hours of wasted developer work. I'm not saying I expected 100% real bugs - maybe even 10% real bugs is worth the hassle. But is it 10%, or 0%? I'm not familiar enough with the statistics which is why I'm asking.

> 
> We can even tolerate expected core dumps when we know the test may crash scylla (like when running the scrub compaction tests that may cause crashes with random sstable injected corruption), but in general I'd rater dtests always fail if a core dump is detected.

Crashes are a bit different from error logs. We expect to **never** have crashes, but we do expect error logs of various types, and have often debated whether a certain "normal" event (e.g., one of the nodes being down, or in the present cases - problems during shutdown) should be reported as an error or warning. This is why it's not obvious that tests need to care exactly what error messages appear in the log - just like they don't care which warning messages appear in the log.

psarna commented:
v4:
 - added logging exceptions from the distributed updater initialization during shutdown, which is otherwise ignored

psarna commented:
Oh, queued already. If it gets through, I'll add the last logging part as a follow-up.

bhalevy commented:
@nyh
> The most important question in this regard of these error-message-failures, how many were real bugs detected, and how many were false positives that caused any hours of wasted developer work. I'm not saying I expected 100% real bugs - maybe even 10% real bugs is worth the hassle. But is it 10%, or 0%? I'm not familiar enough with the statistics which is why I'm asking.

I don't keep up statistics on that.
It happens rarely nowdays that we see new dtest failures due to unexpected errors in the log.
I don't feel it's a waste of my time or anybody elses to examine any such error and vet what to do about it.
If it's benign under the specific test scenario that it can be easily added to the ignore list, and if it indicates a real issue, I bless the day we decided to fail the test on unexpected errors :)

I think that the bottom line here is the dtests function both a regression tests, that we want to pass (and then focusing solely on the functionality will be sufficient), but also as exploration tests in which we piggyback testing the general well being of the system on the regression tests.  That's the reason I also added mandatory cluster shutdown after the test, otherwise we'd never hit issues on shutdown (like this one), that might or might not be directly related to the workload the dtest inflected on the nodes.

nyh commented:
> @nyh
> 
> > The most important question in this regard of these error-message-failures, how many were real bugs detected, and how many were false positives that caused any hours of wasted developer work. I'm not saying I expected 100% real bugs - maybe even 10% real bugs is worth the hassle. But is it 10%, or 0%? I'm not familiar enough with the statistics which is why I'm asking.
> 
> I don't keep up statistics on that.

But is it more than zero times? :-)
Did we ever catch a real bug this way, or was the solution always to change the test to allow another error or "downgrade" the error to a warning?

psarna commented:
The next keeps failing, and the longer I look at the shutdown routine for service levels, the less sense it makes. I'm going to dequeue this fix in hopes to unblock next (although the code in master looks broken too), and instead simplify the shutdown procedure as much as possible.

psarna commented:
v4:
 - got rid of nullifying the `_sl_data_accessor` pointer, since there's no point in doing so on shutdown and there's a potential risk of dereferencing it in a stray continuation
 - got rid of `_distributed_data_updater` future altogether by ensuring that the updates are always initiated on shard0 - before, the function could be called on any shard, but it always invoked a lambda on shard0, but at the same time the function signature did not return a future, which was why the future was kept in an additional attribute (`_distributed_data_updater`). Once we just ensure that this updates can only be initiated on shard0, no additional futures need to be kept and waited for, which simplifies the shutdown procedure.

I'm still not able to reproduce this failure (neither on next nor master), but I still suspect that the master branch is also affected by this, which also makes /next flaky.

psarna commented on service/qos/service_level_controller.cc:
```diff
@@ -103,63 +107,65 @@ future<> service_level_controller::update_service_levels_from_distributed_data()
 
-    return container().invoke_on(global_controller, [] (service_level_controller& sl_controller) {
-        return with_semaphore(sl_controller._global_controller_db->notifications_serializer, 1, [&sl_controller] () {
-            return async([&sl_controller] () {
```
Note to self: this thread is created every time an update is needed (in a seastar::repeat loop), so perhaps it's better off as a coroutine. Alternatively, the whole loop can be moved to a thread/coroutine

psarna replied:
2c4979fb59197fd175bec054eafe505c45f62335, but I'll only post it after this PR (or some future iteration of it) gets merged and next is fixed

psarna commented:
@scylladb/scylla-maint I'm pretty sure that this pull request has a chance of unblocking next (which is in fact a blocked master which luckily passed gating)

psarna commented:
I also tested in debug mode just in case, but no problems were detected

nyh reviewed (approved):
I see Avi has already approved this version. I'll commit it, yet again, hoping that this time it won't fail...

psarna commented:
Still fails... I need to somehow find a way to reliably reproduce this one locally, I'm out of ideas what goes wrong and the core dumps weren't helpful last time. Maybe these new ones at least have some useful context

psarna commented:
And I also hope it's not a coroutine issue again... Maybe I'll also try reproducing the issue with higher inline threshold and coding this one without coroutines if it's that again

psarna commented:
meanwhile, dequeued

bhalevy commented:
what happened to the simple version of the fix?
Was it dequeued too?

psarna commented:
> what happened to the simple version of the fix?
> Was it dequeued too?

Which simple version? No version made it through /next so far, but all of them were also flawed (discovered by closer code inspection later). I hoped the last one is correct because I spent much more time on analyzing it, but I still wasn't able to see this failure locally even once, which makes it harder to reliably check if the fix works or not.

bhalevy commented:
Can we start with something simple like https://github.com/scylladb/scylla/commit/baa91dc07acb4a61c27951c0e706e4d5f9da6cf0
plus warnings about exceptions from _distributed_data_updater that are caught and ignored?

psarna commented:
> Can we start with something simple like [baa91dc](https://github.com/scylladb/scylla/commit/baa91dc07acb4a61c27951c0e706e4d5f9da6cf0)
> plus warnings about exceptions from _distributed_data_updater that are caught and ignored?
We did - and the tests still sometimes segfault with "Stopping service level controller" as the last entry in logs, 
 but at the same time the core dump doesn't have any meaningful backtrace - no segfault or assert, all threads look like idle scylla polling for more work.

psarna commented:
I'm also unable to locally verify my theory about coroutine problems with elevaled inline threshold, because with the original inline threshold setting (2500) row_level.cc never finishes compiling for me.

bhalevy commented:
`service_level_controller::stop` doesn't have to be a coroutine, does it?

psarna commented:
Yes, after the simplification it doesn't anymore. That will be my next attempt, if my current efforts with gdb fail to pinpoint the problem

bhalevy commented:
I tried this:
```
SEASTAR_TEST_CASE(test_exchange_preemption) {
    bool x = false;
    unsigned preempted = 0;
    auto f = later().then([&x] {
            x = true;
        });

    // try to preempt 1000 times. 1 should be enough if not for
    // task queue shaffling in debug mode which may cause co-routine
    // continuation to run first.
    while(preempted < 1000 && !x) {
        preempted += need_preempt(); 
        co_await make_ready_future<>();
    }
    auto save_x = x;
    // wait for later() to complete
    co_await std::exchange(f, make_ready_future<>());
    BOOST_REQUIRE(save_x);
    co_return;
}
```

```
bhalevy@dt seastar$ mode=debug; ninja -C build/$mode test_unit_coroutines && build/$mode/tests/unit/coroutines_test -l all -t test_exchange_preemption
ninja: Entering directory `build/debug'
[91/91] Linking CXX executable tests/unit/coroutines_test
WARNING: debug mode. Not for benchmarking or production
random-seed=4131432382
Running 1 test case...
Entering test module "../../tests/unit/coroutines_test.cc"
../../tests/unit/coroutines_test.cc(0): Test case "test_simple_coroutines" is skipped because disabled
../../tests/unit/coroutines_test.cc(0): Test case "test_abandond_coroutine" is skipped because disabled
../../tests/unit/coroutines_test.cc(0): Test case "test_scheduling_group" is skipped because disabled
../../tests/unit/coroutines_test.cc(0): Test case "test_preemption" is skipped because disabled
../../tests/unit/coroutines_test.cc(0): Entering test case "test_exchange_preemption"
../../src/testing/seastar_test.cc(43): info: check true has passed
../../tests/unit/coroutines_test.cc(191): info: check save_x has passed
=================================================================
==142224==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7fa286d8d888 at pc 0x0000004482e4 bp 0x7fa286d8d5c0 sp 0x7fa286d8d5b8
READ of size 8 at 0x7fa286d8d888 thread T1
    #0 0x4482e3 in seastar::future_state_base::any::has_result() const ../../include/seastar/core/future.hh:503
    #1 0x453703 in seastar::future_state<seastar::internal::monostate>::clear() ../../include/seastar/core/future.hh:606
    #2 0x4496ed in seastar::future_state<seastar::internal::monostate>::~future_state() ../../include/seastar/core/future.hh:614
    #3 0x4496ed in seastar::future<void>::~future() ../../include/seastar/core/future.hh:1337
    #4 0x43f25b in test_exchange_preemption::run_test_case() [clone .actor] ../../tests/unit/coroutines_test.cc:192
    #5 0x44a3cd in std::__n4861::coroutine_handle<void>::resume() const (/home/bhalevy/dev/seastar/build/debug/tests/unit/coroutines_test+0x44a3cd)
    #6 0x44a7b7 in seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() (/home/bhalevy/dev/seastar/build/debug/tests/unit/coroutines_test+0x44a7b7)
    #7 0x6ae815 in seastar::reactor::run_tasks(seastar::reactor::task_queue&) ../../src/core/reactor.cc:2260
    #8 0x6b2ce3 in seastar::reactor::run_some_tasks() ../../src/core/reactor.cc:2669
    #9 0x6b822e in seastar::reactor::run() ../../src/core/reactor.cc:2828
    #10 0x4f076c in seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) ../../src/core/app-template.cc:207
    #11 0x4ee7ca in seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) ../../src/core/app-template.cc:115
    #12 0x47c131 in operator() ../../src/testing/test_runner.cc:70
    #13 0x487403 in __invoke_impl<void, seastar::testing::test_runner::start(int, char**)::<lambda()>&> /usr/include/c++/10/bits/invoke.h:60
    #14 0x485757 in __invoke_r<void, seastar::testing::test_runner::start(int, char**)::<lambda()>&> /usr/include/c++/10/bits/invoke.h:110
    #15 0x483a1f in _M_invoke /usr/include/c++/10/bits/std_function.h:291
    #16 0x50fe7c in std::function<void ()>::operator()() const /usr/include/c++/10/bits/std_function.h:622
    #17 0x657bf6 in seastar::posix_thread::start_routine(void*) ../../src/core/posix.cc:60
    #18 0x7fa28b5163f8 in start_thread (/lib64/libpthread.so.0+0x93f8)
    #19 0x7fa28a73bb52 in __GI___clone (/lib64/libc.so.6+0x101b52)

Address 0x7fa286d8d888 is located in stack of thread T1 at offset 8 in frame
    #0 0x44a6f5 in seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() (/home/bhalevy/dev/seastar/build/debug/tests/unit/coroutines_test+0x44a6f5)

  This frame has 1 object(s):
    [32, 40) 'handle' (line 110)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
Thread T1 created by T0 here:
    #0 0x7fa28bb22f46 in __interceptor_pthread_create (/lib64/libasan.so.6+0x55f46)
    #1 0x658193 in seastar::posix_thread::posix_thread(seastar::posix_thread::attr, std::function<void ()>) ../../src/core/posix.cc:95
    #2 0x657d72 in seastar::posix_thread::posix_thread(std::function<void ()>) ../../src/core/posix.cc:65
    #3 0x4801ce in make_unique<seastar::posix_thread, seastar::testing::test_runner::start(int, char**)::<lambda()> > /usr/include/c++/10/bits/unique_ptr.h:962
    #4 0x47cb64 in seastar::testing::test_runner::start(int, char**) ../../src/testing/test_runner.cc:105
    #5 0x46c296 in init_unit_test_suite ../../src/testing/entry_point.cc:47
    #6 0x7fa28ba6305c  (/lib64/libboost_unit_test_framework.so.1.73.0+0x2605c)

SUMMARY: AddressSanitizer: stack-buffer-underflow ../../include/seastar/core/future.hh:503 in seastar::future_state_base::any::has_result() const
Shadow bytes around the buggy address:
  0x0ff4d0da9ac0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ff4d0da9ad0: f1 f1 f1 f1 00 f2 f2 f2 00 f2 f2 f2 00 00 f2 f2
  0x0ff4d0da9ae0: f8 f8 f2 f2 f8 f8 f2 f2 f8 f8 f2 f2 f8 f8 f2 f2
  0x0ff4d0da9af0: f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f8 f8 f3 f3 f3 f3
  0x0ff4d0da9b00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0ff4d0da9b10: f1[f1]f1 f1 00 f3 f3 f3 00 00 00 00 00 00 00 00
  0x0ff4d0da9b20: 00 00 00 00 f1 f1 f1 f1 f8 f3 f3 f3 00 00 00 00
  0x0ff4d0da9b30: 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1 01 f2
  0x0ff4d0da9b40: 04 f2 00 f2 f2 f2 00 f2 f2 f2 f8 f2 f2 f2 00 f2
  0x0ff4d0da9b50: f2 f2 00 f2 f2 f2 00 f2 f2 f2 00 f2 f2 f2 00 f3
  0x0ff4d0da9b60: f3 f3 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==142224==ABORTING
```

bhalevy commented:
I think that the compiler release the std::exchange'd future as if it was a temporary.

bhalevy commented:
Edit: the `stack-buffer-underflow` problem above happens in debug mode for me also with the existing `test_preemption` case in seastar's coroutines_test

psarna commented:
I'm shipping a decoroutinized patch in a minute, after I verify locally that it doesn't crash

psarna commented:
v5:
 - got rid of the (now rather unnecessary) coroutine in ::stop()

psarna commented:
Maintainers: please consider shipping along with #8506 in order to make it easier to debug core dumps from Jenkins, if they keep happening

bhalevy commented:
LGTM

nyh commented:
I don't know how to review this patch any more (I didn't see any problems in any of the other 7 versions...), but I guess I'll continue to commit it as long as you continue to take care of the fallout (dequeuing and debugging). So, here we go again...

psarna commented:
Looks like next is failing again, no coroutines :( but at least there will be more debug info to look into.

nyh commented:
Oh my god, I'm going to submit this PR to the Guinness Book of World Records as the "unluckiest PR of all time" :-(
I assume you'll do the honor of dequeuing it? Thanks.

psarna commented:
Customary daily dequeue is done. I'll continue investigating tomorrow, here's a message from Benny that I'm going to leverage:
> FWIW, it looks like the service_level_controller::stop hang on shutdown was just reproduced for me locally with scylla in debug mode with topology_test:TestTopology.crash_during_decommission_test

psarna commented:
Good call, tomorrow I'll split the service level controller shutdown procedure into two: stop the updates early (before drain), because they depend on cql, and stop the rest of the controller later.

psarna commented:
And at the same time I'll try to reproduce the problem with the dtest that Benny found

psarna commented:
Found it! It hangs due to the fact that we have an infinite timeout config set up for internal queries, and the service level updater uses an internal query for updates. Once the cluster starts shutting down, the service level updating thread can hang on this infinite timeout.

bhalevy commented:
> Found it! It hangs due to the fact that we have an infinite timeout config set up for internal queries, and the service level updater uses an internal query for updates. Once the cluster starts shutting down, the service level updating thread can hang on this infinite timeout.

Awesome.  That makes sense.

> Good call, tomorrow I'll split the service level controller shutdown procedure into two: stop the updates early (before drain), because they depend on cql, and stop the rest of the controller later.

Let's call the first stage "drain" and the final stage "stop", as the [design doc](https://docs.google.com/document/d/1GHJqHQHAhr1uJtzDaYjmvCy6eAYyLvnzHCyh5NODMRU/edit?usp=sharing) calls.

psarna commented:
v6:
 - fixed the root cause of all previous hangs (which were disguised as core dumps - does Jenkins do that?) : the internal statements for service level updates used infinite timeouts, which caused them to never finish, which prevented proper shutdown
 - split the original stopping routine to separate drain and stop phases - draining is done while cql is still operable, because the service level updater uses the cql layer underneath

psarna commented:
Tests: `topology_test.py:TestTopology.crash_during_decommission_test` - used to hang before the series, succeeds now

xemul commented on main.cc:
```diff
@@ -1460,6 +1460,10 @@ int main(int ac, char** av) {
+            auto stop_sl_updates = defer_verbose_shutdown("service level controller updates", [] {
+                sl_controller.local().drain().get();
+            });
+
```
If someone does `nodetool drain` and then stops scylla, this deferred action will be called in the condition where commitlog is down and migration manager is stoped, i.e. as if it was called _after_ `drain_on_shutdown`. Is that OK?

psarna replied:
It is now, because the updater only reads from Scylla, it doesn't need to write anything.

xemul replied:
OK, but why is `s_l_controller::drain` put to be run before `storage_service::drain_on_shutdown`?

psarna replied:
Actually there's no reason now. I'll put it after, to match the behavior if somebody runs `nodetool drain`

xemul commented on service/qos/service_level_controller.cc:
```diff
@@ -86,15 +86,26 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
+}
+
+future<> service_level_controller::stop() {
+    return make_ready_future<>();
```
If start aborts due to an exception before `service_level_controller::drain` is deferred, then the whole stop becomes a no-op.

psarna replied:
Right, that's no good

psarna replied:
I guess I should either make draining idempotent and call it in stop() or move the updater initialization right before deferring the drain. I'll go with making it idempotent.

bhalevy commented:
> fixed the root cause of all previous hangs (which were disguised as core dumps - does Jenkins do that?)

ccm forces the core dump is the node does not shut itself down in a timely manner and times out.
In production systemd will do that.

nyh commented:
> > fixed the root cause of all previous hangs (which were disguised as core dumps - does Jenkins do that?)
> 
> ccm forces the core dump is the node does not shut itself down in a timely manner and times out.
> In production systemd will do that.

I didn't know that... That's confusing. Could it at least write something to the log saying that it's doing that?

xemul commented on main.cc:
```diff
@@ -1452,6 +1452,10 @@ int main(int ac, char** av) {
+            auto stop_sl_updates = defer_verbose_shutdown("service level controller updates", [] {
+                sl_controller.local().drain().get();
+            });
+
```
Does it make sense to have separate action for `s_l_controller::drain` now? It's immediately followed by `sharded<s_l_controller>::stop`, so if you remove this draining defer, seastar will call `s_l_controller::stop` automagically and this `stop` already calls `s_l_controller::drain` inside.

psarna replied:
Yeah, this deferrer can be reduced now

bhalevy commented:
> > > fixed the root cause of all previous hangs (which were disguised as core dumps - does Jenkins do that?)
> > 
> > 
> > ccm forces the core dump is the node does not shut itself down in a timely manner and times out.
> > In production systemd will do that.
> 
> I didn't know that... That's confusing. Could it at least write something to the log saying that it's doing that?

It does, though it's hard to find.
It should be written to dtest.log, but it doesn't have the test name prefix since it comes from the ccm layer rather than the dtest layer.  They can be matched together using the PID column that is the same for both dtest and log ccm messages in the same nosetests process.

psarna commented:
v7:
 - made drain() idempotent
 - ensured that drain is always called before stop
 - removed the (now redundant) separate deferred action for draining the service level controller

psarna commented:
ping

avikivity commented on service/qos/service_level_controller.cc:
```diff
@@ -86,15 +86,30 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
+        return make_ready_future<>();
     }
-    return std::exchange(_distributed_data_updater, make_ready_future<>());
+    _drained = true;
```
Could two drains be called in parallel? In that case the second will return early.

If that's a problem, we can use a shared_future to indicate when draining is complete to all callers.

psarna replied:
Currently the only user of `service_level_controller::drain` is `service_level_controller::stop`, which is only called once on shutdown, so two drains would not be called in parallel.

avikivity replied:
Ok. So I don't see why you needed to extract drain().

psarna replied:
It turned out I didn't. It's a byproduct of earlier iterations, but it's good practice anyway (because the update process can have virtual executors and depend on multiple services), so I let it stay. I can also remove it from this series if requested.

psarna replied:
(please let me know if it *is* requested)

avikivity replied:
No, it's fine as is.

bhalevy replied:
I don't think that can happen, yet it looks like if it will `std::exchange(_global_controller_db->distributed_data_update, make_ready_future<>())` it would be idempotent and won't need to add the `_drained` flag at all.

psarna replied:
As long as `abort_source::request_abort()` and `semaphore::broken()` are also idempotent, but it looks like they are. I'll get rid of the flag.

psarna replied:
Oh wait - semaphore::broken might be, but request_abort isn't, it would have to be wrapped in a check. Still, it's easy.

bhalevy replied:
You're right about `request_abort`, sorry.  And indeed `abort_requested()` can come to the rescue.

avikivity commented:
@bhalevy @nyh are you okay with this?

psarna commented:
v8:
 - got rid of the explicit `_drained` variable by ensuring that the whole drain operation is idempotent anyway

bhalevy reviewed (approved):
LGTM

nyh commented:
Committed. Hopefully 10th time's a charm :-)

psarna commented:
Sighhhhhhhhhhhhhhhhhhhhhhh....

psarna commented:
I'll dequeue - everything is fine this time, except that a test which tries unrealistic allocation sizes messed with the address sanitizer from debug mode. I'll provide fixed version 10000000 soon, and meanwhile also dequeue.

psarna commented:
Oh wait!!! It's not this series. It's the one with rapidjson allocators.
